packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,500 @@
+            if envelope is None:
+                await asyncio.sleep(sleep)
+                timeout_count += sleep
+            else:
+                not_received = False
+        if not_received:
+            self.timed_out_request_ids.add(request_id)
+        return envelope
+
+    def build_envelope(self, request: Request) -> Envelope:
+        """
+        Process incoming API request by packaging into Envelope and sending it in-queue.
+
+        The Envelope's message body contains the "performative", "path", "params", and "payload".
+
+        :param http_method: the http method
+        :param url: the url
+        :param param: the parameter
+        :param body: the body
+        """
+        client_id = uuid4().hex
+        uri = URI(request.url)
+        context = EnvelopeContext(connection_id=self.connection_id, uri=uri)
+        # TODO: replace with HTTP protocol
+        msg = {
+            "performative": request.method,
+            "path": request.path,
+            "query": request.query,
+            "payload": request.body,
+        }
+        msg_bytes = json.dumps(msg).encode()
+        envelope = Envelope(
+            to=self.address,
+            sender=client_id,
+            protocol_id=PublicId.from_str("fetchai/http:0.1.0"),
+            context=context,
+            message=msg_bytes,
+        )
+        return envelope
+
+    def build_response(self, envelope: Optional[Envelope] = None) -> Response:
+        if envelope is not None:
+            # Response Envelope's msg will be a JSON with 'status_code', 'response', and 'payload' keys
+            resp_msg = json.loads(envelope.message.decode())
+            response = Response(resp_msg["status_code"], resp_msg["message"])
+        else:
+            response = Response(REQUEST_TIMEOUT, b"Request Timeout")
+        return response
+
+
+def HTTPHandlerFactory(channel: HTTPChannel):
+    class HTTPHandler(BaseHTTPRequestHandler):
+        """HTTP Handler class to deal with incoming requests."""
+
+        def __init__(self, *args, **kwargs):
+            """Initialize a HTTP Handler."""
+            self._channel = channel
+            super(HTTPHandler, self).__init__(*args, **kwargs)
+
+        @property
+        def channel(self) -> HTTPChannel:
+            """Get the http channel."""
+            return self._channel
+
+        def get_url(self) -> str:
+            """Get url."""
+            url = "http://{}:{}{}".format(*self.server.server_address, self.path)
+            return url
+
+        def do_HEAD(self):
+            """Deal with header only requests."""
+            self.send_response(SUCCESS)
+            self.send_header("Content-Type", "text/plain; charset=utf-8")
+            self.end_headers()
+
+        def do_GET(self):
+            """Respond to a GET request."""
+            request = self.build_request("get")
+
+            response = self.channel.loop.run_until_complete(
```

Comment:
because in the tests you are not using the connection via the `Multiplexer`, so `self.channel.loop` is not running.
If it is not running, you can use `run_until_complete`, but if it is already running (which is the case when you run the connection with the Multiplexer) you'll get an error!

Reply:
https://github.com/fetchai/agents-aea/pull/901